storage/src/main/java/zipkin2/storage/kafka/KafkaStoreServerSupplier.java

```diff
@@ -0,0 +1,328 @@
+import zipkin2.storage.QueryRequest;
+import zipkin2.storage.kafka.internal.Traces;
+
+import static zipkin2.storage.kafka.streams.DependencyStoreTopologySupplier.DEPENDENCIES_STORE_NAME;
+import static zipkin2.storage.kafka.streams.TraceStoreTopologySupplier.AUTOCOMPLETE_TAGS_STORE_NAME;
+import static zipkin2.storage.kafka.streams.TraceStoreTopologySupplier.REMOTE_SERVICE_NAMES_STORE_NAME;
+import static zipkin2.storage.kafka.streams.TraceStoreTopologySupplier.SERVICE_NAMES_STORE_NAME;
+import static zipkin2.storage.kafka.streams.TraceStoreTopologySupplier.SPAN_IDS_BY_TS_STORE_NAME;
+import static zipkin2.storage.kafka.streams.TraceStoreTopologySupplier.SPAN_NAMES_STORE_NAME;
+import static zipkin2.storage.kafka.streams.TraceStoreTopologySupplier.TRACES_STORE_NAME;
+
+/**
+ * Server to enable access to local stores.
+ * <p>
+ * Given the partitioned nature of local stores, a RPC layer is required to allow accessing
+ * distributed state. This component exposes access to local state via Http call from {@link
+ * KafkaSpanStore}
+ *
+ * @since 0.6.0
+ */
+public class KafkaStoreServerSupplier implements Supplier<Server> {
+  static final Logger LOG = LoggerFactory.getLogger(KafkaStoreServerSupplier.class);
+  static final Gson GSON = new Gson();
+
+  final KafkaStorage storage;
+  final long minTracesStored;
+  final int httpPort;
+  final PrometheusMeterRegistry prometheusRegistry;
+
+  KafkaStoreServerSupplier(KafkaStorage storage) {
+    this.storage = storage;
+
+    this.minTracesStored = storage.minTracesStored;
+    this.httpPort = storage.httpPort;
+    this.prometheusRegistry = storage.prometheusRegistry;
+  }
+
+  @Override public Server get() {
+    ServerBuilder builder = new ServerBuilder();
+    builder.http(httpPort);
+    // Metadata-related services
+    builder.service("/instances", getInstances());
+    builder.service("/instances/:store_name", getInstancesByStore());
+    // Traces-related services
+    builder.annotatedService(getTraces()); // kv.range() + foreach(kv.get()).filter()
+    builder.service("/traces/:trace_id", getTrace()); // kv.get()
+    // Service names related
+    builder.service("/service-names", getServiceNames()); // kv.all()
+    builder.service("/service-names/:service_name/span-names",
+        getSpanNamesByServiceName()); // kv.get()
+    builder.service("/service-names/:service_name/remote-service-names",
+        getRemoteServiceNamesByServiceName()); // kv.get()
+    builder.annotatedService(getDependencies());
+    builder.service("/autocomplete-tags", getAutocompleteTagKeys());
+    builder.service("/autocomplete-tags/:key", getAutocompleteTagValues());
+    builder.service("/metrics", getMetrics());
+    return builder.build();
+  }
+
+  private Service<HttpRequest, HttpResponse> getMetrics() {
+    return (ctx, req) -> HttpResponse.of(MediaType.PLAIN_TEXT_UTF_8, prometheusRegistry.scrape());
+  }
+
+  private Service<HttpRequest, HttpResponse> getAutocompleteTagValues() {
```

Comment:
You might want to use an annotated service, and make sure every method returns `AggregatedHttpResponse`.

http://line.github.io/armeria/server-annotated-service.html

As currently implemented, every service is doing blocking I/O on the event loop so this server will either perform very poorly or deadlock. Annotated service methods that return `AggregatedHttpResponse` automatically run on a threadpool dedicated to blocking logic and won't have the problem.